### PR TITLE
CLDR-16531 BRS: v44 GenerateLanguageContainment

### PR DESCRIPTION
**Pull request by srl295**
CLDR-16531

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

Added some loop detection and avoidance, CLDR-15020

### CODE REVIEW

**Comment by srl295** [comment]
ping?

File: common/supplemental/languageGroup.xml [hunk 1]
```diff
@@ -9,32 +9,32 @@ CLDR data files are interpreted according to the LDML specification (http://unic
...
...jmr jms jni juh juk juo juw kaj kbp kcc kcg kcj kdh kdm kdx kdz kef ken kes keu kez kfl kfn kgt khj kia kid kkm kku klc klk klo kma kmi kmy knf knp kou kov kow kph kpk kpl kpo kqg kqk kqm kqs krh kro krp krx ksa kss kub kvm kxb kye kza kzc kzr lan lar las ldg ldh ldm ldq lee lef lgq lia lip lma lmp lns lob lof lro luq mae mbv mcj mcs mcu mdd mea mfc mfd mff mfn mfq mfv mgj mgn mgo mhk mij mkl mko mlo moj mor mos mql mru msj msw mtk mua muc muo mxl myg myj myk mzk mzm mzv mzw naj nat naw nbm nbo nbp nbr nbv nbw ncr ncu ndd ndi ndr ndt ndv ndz ned nfr nfu nga ngb nge ngg nhu nie nin njr nko nkq nku nkz nlu nmj nmr nmz nnu nnw noy nqk nsc nti ntm nuh nup nuv nyb nza nzi nzz obu oda odu ofu ogb ogc ogg ogo ogu okb oke oku opa orx otr pbl pbo pbp pcn pil pmb png pnl pnq pny pnz pug pwb pym res ruy saf sav scv sde sef sev sfw sg sha shw shz sig sil snf snj snw soy spp sqm srr ssl swf sxw tbm tbz tcd tde tdl tem tfi tgw tgy thy tik tiq tiv tkq tnr toz tsp tsw ttb tug tui tuz tvd twn uda uiv ukq ukw ula umm urh usk uta uth utr uya vag vit vor vut wbf wci wem wib wlx wo wof wom wrn wss wwa www xab xkb xoc xrb xsm xsn xwe xwl yam yav yay yaz yba ybj ybl yer yky yng ynq yo zna</languageGroup>
```
Comment (markusicu) [inline common/supplemental/languageGroup.xml:14]: FYI GitHub doesn't show what's changing on some of these lines.

File: common/supplemental/languageGroup.xml [hunk 2]
```diff
@@ -9,32 +9,32 @@ CLDR data files are interpreted according to the LDML specification (http://unic
...
+		<languageGroup parent="ber">aao aeb arq ary auj ayl gha gnc jbn kab mt nxm siz sqr swn taq tez thv thz tmh xaa zen zgh</languageGroup>
```
Comment (markusicu) [inline common/supplemental/languageGroup.xml:27]: Half of these languages are different. This seems like a surprisingly large change...?

File: common/supplemental/languageGroup.xml [hunk 3]
```diff
@@ -83,38 +83,38 @@ CLDR data files are interpreted according to the LDML specification (http://unic
...
 		<languageGroup parent="plf">aaz alo alp amq amv aoz apx asl baj bay bgy bkx bvt bzn emw frd gal hbu heg hrk hti huk huw idt ilu jal jmd kei kem kje krd kuv kvp kvv kwh kzb kzd kzx lcd lcl lcq lex lhh lka llg lox lti lva mgm mhs mss mxz nae nil nni nxa nxl oni plh rgu row sgu skz spb spr srw sve sws szn tdt tet tev tlt tlu tlv tre tve twu txn txq tzn udj urn vme weo wet wha wmh</languageGroup>
-		<languageGroup parent="poz">adr akg aol bdb beg bhp bhr bhv bmm bny buc bzc day dro dun duq duw end hvn ila kdg kod ksx kuk kvh kxn kyd kyi kys kzi lbx lgi ljl llk lmq lmy lur lwt mel mg mhy mnb mqy msh mvd mxr nea nfa nij nrm nrz nsy nxe nxg omy otd pez phi pku ple plf pne pnm pqe pqw puc puf reb ree ror sau sdx sge skg ski snv spg ssq sya tdx tjg tkg ttg ttw twy txy ulu umi wew whk wnk xbr xkd xkl xky xmv xmw xxk zbc zbe zbw</languageGroup>
+		<languageGroup parent="poz">adr akg aol bdb beg bhp bny day dro end hvn ila kdg kod ksx kuk kvh kxn kyd kyi kys kzi lgi ljl llk lmq lmy lur lwt mel mnb mqy mvd mxr nea nfa nrm nrz nsy nxe nxg pez phi ple plf pne pnm pqe pqw puc puf reb ree ror sau sdx sge ski snv spg ssq ttg ttw ulu umi wew whk wnk xbr xkd xkl xky xxk zbc zbe zbw</languageGroup>
```
Comment (markusicu) [inline common/supplemental/languageGroup.xml:86]: Another large number of changes here.

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/Containment.java [hunk 4]
```diff
@@ -203,16 +204,41 @@ public Set<String> getSubontinents() {
...
     }
 
     private static void getAllDirected(
             Multimap<String, String> multimap,
             String lang,
             ArrayList<String> target,
-            Set<List<String>> targets) {
-        target.add(lang);
+            Set<List<String>> targets,
+            List<String> depth,
```
Comment (markusicu) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/Containment.java:222]: "depth" sounds like an integer. Maybe "ancestors"?

File: tools/cldr-rdf/external/RawLanguageContainment.txt
Comment (markusicu) [inline tools/cldr-rdf/external/RawLanguageContainment.txt:1]: The files in this folder look like inputs -- exported from some database, scraped from wikidata, or something. Not sure how I would review them.

**Review by markusicu — COMMENTED** [review]
Not sure I can do this review justice. I haven't worked with these files.

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/Containment.java (hunk 4 above)
Comment (srl295) [reply tools/cldr-code/src/main/java/org/unicode/cldr/util/Containment.java:222]: That's a good point.  It was originally an integer but I changed it when I needed more than just the depth.  Will update.

File: tools/cldr-rdf/external/RawLanguageContainment.txt
Comment (srl295) [reply tools/cldr-rdf/external/RawLanguageContainment.txt:1]: Yes they are updated based on changes in wiki data. The scraped files are checked in for the ability to work with them again even if the external web went away or changed dramatically

**Review by macchiati — APPROVED** [review]

**Review by macchiati — CHANGES_REQUESTED** [review]
I looked this over a bit more, and we shouldn't merge. I'll work on it next week. (This is not a problem with Steven's changes: there is something else in the original code.)

**Comment by srl295** [comment]
> I looked this over a bit more, and we shouldn't merge. I'll work on it next week. (This is not a problem with Steven's changes: there is something else in the original code.)

Thanks mark

File: common/supplemental/languageGroup.xml (hunk 1 above)
Comment (FrankYFTang) [inline common/supplemental/languageGroup.xml:14]: the change is add ffm

**Pull request closed without merging**